File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def bfs(maps, n, m, x, y):
    queue = deque([[x, y]])
    
    while queue:
        curX, curY = queue.popleft()
        
        for nextX, nextY in [[curX + 1, curY], [curX, curY + 1], [curX - 1, curY], [curX, curY - 1]]:
            if nextX < 0 or nextX >= n or nextY < 0 or nextY >= m:
                continue
            
            if maps[nextX][nextY] == 1:
                maps[nextX][nextY] = 1 + maps[curX][curY]
                queue.append([nextX, nextY])
    
    
def solution(maps):
    n, m = len(maps), len(maps[0])
    
    bfs(maps, n, m, 0, 0)
    
    answer = maps[n - 1][m - 1]
    
    if answer == 1:
        return -1
    
    return answer
    

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
	let N = maps.length;
	let M = maps[0].length;

	let distance = Array.from({ length: N }, () => Array(M).fill(-1));
	distance[0][0] = 1;

	let dx = [-1, 0, 1, 0];
	let dy = [0, -1, 0, 1];

	let queue = [[0, 0]];
	while (queue.length) {
		let [curX, curY] = queue.shift();

		for (let i = 0; i < 4; i++) {
			let nextX = curX + dx[i];
			let nextY = curY + dy[i];

			if (
				nextX < 0 ||
				nextX >= N ||
				nextY < 0 ||
				nextY >= M ||
				maps[nextX][nextY] === 0
			)
				continue;

			if (distance[nextX][nextY] === -1) {
				distance[nextX][nextY] = distance[curX][curY] + 1;
				queue.push([nextX, nextY]);
			}
		}
	}

	return distance[N - 1][M - 1];
}